# Corrélation entre les features

## 1)Import des données

In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
df = pd.read_csv("../6.Data/kaggle_b2_fraud_train_v4.csv")

In [2]:
df

,customer_id,account_id,age,tenure_months,annual_income_eur,credit_score,num_transactions_30d,avg_amount_30d_eur,max_amount_30d_eur,days_since_last_login,...,internal_signal_2,internal_signal_3,internal_signal_4,internal_signal_5,internal_signal_6,internal_signal_7,internal_signal_8,terms_accepted_flag,partner_risk_indicator,target_is_fraud
0,CUST_6O9Q8D4I36,ACC_TXXXTNEUVKFY,34,108,38635.01,544,20,60.92,80.16,4.9,...,2.77585,-0.35668,0.73256,-0.90929,0.38747,-0.21200,1.78510,1,NaN,0
1,CUST_FGUGTW230C,ACC_70VD7A4FFWCW,48,2,19912.97,703,21,112.11,571.12,0.3,...,-0.65897,0.50084,0.23241,-1.16817,-0.82509,-0.78065,-1.25603,1,2.244073,0
2,CUST_8ZI3LCBZ0W,ACC_AF53381QSC0L,27,0,20326.87,720,25,73.61,492.57,4.6,...,0.68621,-1.30771,-0.91109,0.92254,-0.63157,0.89172,-0.34545,1,NaN,1
3,CUST_5MP3AR41CJ,ACC_U7WZGJ486LIV,45,49,38452.47,703,17,47.53,204.18,25.3,...,-1.95970,-1.68684,0.85302,1.28198,1.06118,0.99292,-0.05549,1,NaN,0
4,CUST_GNPL83JB0J,ACC_XW7DS3ED5J4Y,37,46,23065.56,594,13,99.95,734.09,12.8,...,0.32109,-1.51914,-0.27024,-0.61402,1.32524,0.78992,-0.96780,1,NaN,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
159995,CUST_I81IW5SVRQ,ACC_UPDTFTYTSM0A,56,0,34775.62,727,21,51.72,226.11,3.8,...,0.91351,-1.06238,-0.95945,-0.66641,-0.70669,-0.25664,-1.02678,1,NaN,0
159996,CUST_QT6DDEMKTJ,ACC_97NE0LBL5W9U,41,4,88617.57,770,18,42.92,171.07,15.1,...,2.05022,-1.10418,-0.89959,0.45220,0.00000,0.58802,0.95938,1,NaN,1
159997,CUST_I0JS1GTS98,ACC_9JJ84W64Z7GX,30,2,41148.54,738,20,29.34,119.81,0.7,...,-1.56646,0.16181,0.12341,-1.06187,0.30498,-0.07834,0.33083,1,NaN,0
159998,CUST_L7GUCJ3TFY,ACC_NGFXDR7HW1ZS,56,6,15558.62,719,25,88.56,553.16,22.6,...,0.67588,0.14722,-0.40554,-0.11118,0.85848,1.69012,0.06194,1,-1.132891,0


## 2) Observation

In [6]:
import pandas as pd
import numpy as np

# =========================
# 1️⃣ Sélection des variables numériques
# =========================

df_numeric = df.select_dtypes(include=["int64", "float64"])

# =========================
# 2️⃣ Matrice de corrélation (valeur absolue)
# =========================

corr_matrix = df_numeric.corr().abs()

# =========================
# 3️⃣ On garde uniquement la partie supérieure
# (évite les doublons A,B et B,A)
# =========================

mask = np.triu(np.ones(corr_matrix.shape), k=1).astype(bool)
upper = corr_matrix.where(mask)

# =========================
# 4️⃣ On récupère les 20 plus fortes corrélations
# =========================

top_pairs = (
    upper.unstack()
    .dropna()
    .sort_values(ascending=False)
    .head(20)
)

# =========================
# 5️⃣ Affichage
# =========================

print("Top 20 des plus fortes corrélations :")
print(top_pairs)

Top 20 des plus fortes corrélations :
max_amount_30d_eur      avg_amount_30d_eur        0.790711
target_is_fraud         ip_risk_z                 0.258598
                        device_trust_z            0.164562
                        is_vpn                    0.145544
                        is_new_device             0.142144
                        tenure_months             0.086804
                        chargebacks_12m           0.078046
                        failed_payments_6m        0.074573
partner_risk_indicator  is_new_device             0.025308
target_is_fraud         partner_risk_indicator    0.023133
partner_risk_indicator  avg_amount_30d_eur        0.022289
                        max_amount_30d_eur        0.021860
                        tenure_months             0.021631
                        internal_signal_6         0.017922
                        internal_signal_3         0.017095
                        is_vpn                    0.017025
                  

In [7]:
import pandas as pd
import numpy as np

# On garde uniquement les variables numériques
df_numeric = df.select_dtypes(include=["int64", "float64"])

# Calcul corrélations
corr_with_target = (
    df_numeric.corr()["target_is_fraud"]
    .drop("target_is_fraud")
)

# Tri par importance (valeur absolue)
corr_sorted = corr_with_target.reindex(
    corr_with_target.abs().sort_values(ascending=False).index
)

print(corr_sorted)


ip_risk_z                 0.258598
device_trust_z           -0.164562
is_vpn                    0.145544
is_new_device             0.142144
tenure_months            -0.086804
chargebacks_12m           0.078046
failed_payments_6m        0.074573
partner_risk_indicator    0.023133
max_amount_30d_eur       -0.004194
support_tickets_90d       0.003761
internal_signal_2         0.003761
num_devices_30d          -0.002646
age                      -0.002620
num_transactions_30d      0.002596
annual_income_eur        -0.002528
internal_signal_6         0.002305
internal_signal_4        -0.001898
internal_signal_7        -0.001294
avg_amount_30d_eur       -0.001265
days_since_last_login     0.001129
internal_signal_8        -0.001099
credit_score              0.000971
internal_signal_3         0.000794
internal_signal_1         0.000692
internal_signal_5         0.000483
terms_accepted_flag            NaN
Name: target_is_fraud, dtype: float64


Les corrélations entre chargeback_resolution_time_days et post_event_status_code avec target_is_fraud sont trop corrélés. Ceci s'explique par le fait que ces deux colonnes ne prédisent pas la fraud, elles sont des conséquences de celle-ci. On supprime alors ces deux colonnes.

Pour les colonnes corrélés > 0,85, on les supprime pour une regression logistique.
Pour Random Forest on peut garder les colonnes qui sont < 0,95.

| colonne(s) concernée(s)                              | explication                                                                                           | action                                                                 |
|-------------------------------------------------------|--------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------|
| chargeback_resolution_time_days                      | Variable post-événement : conséquence directe d’une fraude (data leakage)                            | Supprimer la colonne                                                  |
| post_event_status_code                               | Variable générée après investigation / traitement (data leakage)                                     | Supprimer la colonne                                                  |
| Variables avec corrélation > 0.85 (entre features)   | Forte multicolinéarité pouvant déstabiliser une régression logistique                                | Supprimer une des variables corrélées                                 |
| Variables avec corrélation > 0.95 (entre features)   | Corrélation quasi parfaite (information redondante)                                                  | Supprimer une des variables corrélées                                 |
| Variables avec corrélation < 0.95 (Random Forest)    | Les modèles basés sur les arbres gèrent mieux la multicolinéarité                                    | Conserver les variables                                               |
